cmd/completion.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/pulumi/pulumi/pkg/util/cmdutil"
+)
+
+// newCompletionCmd returns a new command that, when run, generates a bash or zsh completion script for the CLI.
+// It is hidden by default since it's not commonly used outside of our own build processes.
+func newCompletionCmd(root *cobra.Command) *cobra.Command {
+	return &cobra.Command{
+		Use:    "completion <SHELL>",
```

Comment:
Perhaps we should keep this command name closer to the original - `gen-completion`?

Reply:
fine with me